### ISSUE TITLE
API routes break when written using camelCase and deployed to Vercel

### ISSUE THREAD

**Issue by Mi-Ski** [vercel/next.js#46126]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

 Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.14.2
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      13.1.7-canary.21
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

Middleware / Edge (API routes, runtime)

### Link to the code that reproduces this issue

https://github.com/Mi-Ski/dipper/commit/bac763d54f4a5b94bc43cf1f504f2bd1ab631858#diff-07141191bd8fbeba4774631872da5b2761116feb4207d0d37e782c403f670ab7

### To Reproduce

<img width="383" alt="image" src="https://user-images.githubusercontent.com/63957349/219975744-cba0d059-ad7c-4d8c-b0eb-2ce3d0c54ed7.png">

Steps to reproduce:
1) Use or change exisitng /api/ route file name so that it contains an uppercase letter.
2) Build the project (l deployed it on Vercel)

### Describe the Bug

Api route breaks when it contains uppercase letter. This happens when the project is built and deployed on Vercel. 

"Application error: a client-side exception has occurred (see the browser console for more information)."  is not a helpful error message pointing to what caused the bug. There is no mention on https://nextjs.org/docs/api-routes that api routes have to be written exclusively in lowercase.

Server breaks only when built and deployed, camelCase api routes work fine on localhost, which just adds to confusion.

### Expected Behavior

Server respects the camelCase convention of naming javascrip files and api route paths containing uppercase letters work as expected.

### Which browser are you using? (if relevant)

Chrome 110.0.5481.104

### How are you deploying your application? (if relevant)

Vercel

**Comment by icyJoseph** [comment]
Hi,

The problem here is that git and case-sensitivity are tricky.

In that branch when I look at your API routes:

![Screenshot 2023-02-20 at 08 47 14](https://user-images.githubusercontent.com/21013447/220044049-3bace8fb-c601-47d2-a86d-bed1034f0ce2.png)

There's no `getAll` route. So `git` did not pick up on the change.

You can use this SO answer to get around the case sensitivity issue, https://stackoverflow.com/a/11183844:

It basically means that you rename your file `getall.js` to something else, like `tmp.js`, and then rename it back to `getAll.js`, all using the `git mv` command.

## Client side error not helpful

This is not related to your root issue but here:

```js
 useEffect(() => {
    (async () => {
      // You really need to wrap this with a try/catch
      const getUser = await fetch("/api/user"); // this can throw
      const getUserJson = await getUser.json(); // this can throw
      setUser(getUserJson);

      const tweets = await fetch("/api/tweets/getAll"); // this can throw
      const tweetsjson = await tweets.json();  // this can throw
      setPosts(tweetsjson.documents);

      setLoading(false);
      console.log("useeffect");
    })();
  }, [setPosts, setUser]);
```

You are doing a lot of asynchronous work without accounting for failure scenarios. I've labelled lines that may fail with `// this can throw`, should any of those cause an issue, your entire handler falls apart, and you'll see some useless error messages.

You should wrap it all on a try/catch, collect the failure reason and set state appropriately.

**Comment by GabenGar** [comment]
Yeah this case-(in)sensitivity of filenames is further exacerbated by windows/unix differences.
And that exacerbated even further by filepath/url differences. What you think is just a file on your computer is actually a string identifier which gets passed through several languages/programs/platforms/servers/networks and constantly flipflops between being a path or a url and even different encodings in worst case scenarios.
Gladly it generally "just works", but when it doesn't it's super annoying to debug and even more annoying to fix.
Basically use only lowercase latin letters and `-` instead whitespaces in your source code pathnames. Or for any path which can potentially be shared.

**Comment by aly-tq** [comment]
<img width="944" alt="image" src="https://github.com/vercel/next.js/assets/144067089/57a7cf11-b430-4a01-a20d-a9c2c77b0a4b">

Hitting this error when clicking the + sign on 

"aly@terralytiq.com, Member"

**Comment by ashikpy** [comment]
This happened to me right now, Nextjs should def should include somewhere in there documentation. this really took a lot of time to debug. **OH GOD**

**Comment by marcoshernanz** [comment]
Thanks for the report. The linked commit does not contain a `getAll` route. It contains `pages/api/tweets/getall.js`, while the client requests `/api/tweets/getAll`. Git did not record the case-only rename on Windows, and the project later fixed the mismatch by changing the request back to `/getall`. A genuinely committed camel-cased API route works in production. Closing since this is a repository casing mismatch, not a Next.js bug.
